### PR TITLE
pythonPackages.weasyprint: init at 45

### PR DESCRIPTION
**Pull request by elohmeier**
###### Motivation for this change
add WeasyPrint to nixpkgs

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/development/python-modules/weasyprint/default.nix [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  pytestrunner,
+  isPy3k
+}:
+
+buildPythonPackage rec {
+  pname = "WeasyPrint";
+  version = "45";
+  disabled = !isPy3k;
+
+  doCheck = false;
```
Comment (dotlambda) [inline pkgs/development/python-modules/weasyprint/default.nix]: Why?

File: pkgs/development/python-modules/weasyprint/default.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    substituteInPlace weasyprint/fonts.py --replace "libfontconfig.so.1" "${fontconfig.lib}/lib/libfontconfig.so.1"
+    substituteInPlace weasyprint/fonts.py --replace "libpangoft2-1.0.so" "${pango.out}/lib/libpangoft2-1.0.so"
+
+    # MacOS
+    substituteInPlace weasyprint/text.py --replace "libgobject-2.0.dylib" "${glib.out}/lib/libgobject-2.0.dylib"
+    substituteInPlace weasyprint/text.py --replace "libpango-1.0.dylib" "${pango.out}/lib/libpango-1.0.dylib"
+    substituteInPlace weasyprint/text.py --replace "libpangocairo-1.0.dylib" "${pango.out}/lib/libpangocairo-1.0.dylib"
+    substituteInPlace weasyprint/fonts.py --replace "libfontconfig-1.dylib" "${fontconfig.lib}/lib/libfontconfig.1.dylib"
+    substituteInPlace weasyprint/fonts.py --replace "libpangoft2-1.0.dylib" "${pango.out}/lib/libpangoft2-1.0.dylib"
+  '';
```
Comment (dotlambda) [inline pkgs/development/python-modules/weasyprint/default.nix]: Please use a proper patch with `substituteAll` and `stdenv.hostPlatform.extensions.sharedLibrary`. Also, no need to specify `.out`.

File: pkgs/development/python-modules/weasyprint/default.nix [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  glib,
+  pango,
+  fontconfig,
+  stdenv,
+  pytestrunner,
+  isPy3k
+}:
+
+buildPythonPackage rec {
+  pname = "WeasyPrint";
```
Comment (dotlambda) [inline pkgs/development/python-modules/weasyprint/default.nix]:
```suggestion
  pname = "weasyprint";
```
You'll need to put `pname = "WeasyPrint";` inside `fetchPypi`.

File: pkgs/development/python-modules/weasyprint/default.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  version = "45";
+  disabled = !isPy3k;
+
+  doCheck = false;
+  propagatedBuildInputs = [ cairosvg pyphen cffi cssselect lxml html5lib tinycss pygobject2 pytestrunner ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/weasyprint/default.nix]: Some of these are only `checkInputs`.

**Review by dotlambda — CHANGES_REQUESTED** [review]

File: pkgs/development/python-modules/weasyprint/default.nix (hunk 1 above)
Comment (elohmeier) [reply pkgs/development/python-modules/weasyprint/default.nix]: I've added the tests now :-)

File: pkgs/development/python-modules/weasyprint/default.nix (hunk 2 above)
Comment (elohmeier) [reply pkgs/development/python-modules/weasyprint/default.nix]: Thanks for that helpful tip, didn't know that way! It looks way cleaner now :-)

File: pkgs/development/python-modules/weasyprint/default.nix (hunk 3 above)
Comment (elohmeier) [reply pkgs/development/python-modules/weasyprint/default.nix]: Done

File: pkgs/development/python-modules/weasyprint/default.nix (hunk 4 above)
Comment (elohmeier) [reply pkgs/development/python-modules/weasyprint/default.nix]: I've moved them.

File: pkgs/development/python-modules/weasyprint/default.nix [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  # ignore failing pytest
+  checkPhase = "pytest -k 'not test_font_stretch'";
+
+  # ignore failing flake8-test
+  prePatch = ''
+    substituteInPlace setup.cfg \
+        --replace '[tool:pytest]' '[tool:pytest]\nflake8-ignore = E501'
+  '';
+
+  checkInputs = [ pytestrunner pytest-isort pytest-flake8 pytestcov ];
```
Comment (dotlambda) [inline pkgs/development/python-modules/weasyprint/default.nix]: Please add `pytest` here too. I'll remove it from `pytest-isort`'s `propagatedBuildInputs` in the future.

Comment (dotlambda) [inline pkgs/development/python-modules/weasyprint/default.nix]: Please add pytest here. I'll remove it from pytest-isort's propagatedBuildInputs in the future.

**Comment by dotlambda** [comment]
We'll have to wait until #56839 is merged.

**Comment by risicle** [comment]
#56839 is merged, but for me now on master, the `python3.7-CairoSVG-2.3.0` dependency fails its `isort-check` test. Non-nixos linux x86_64.

**Comment by elohmeier** [comment]
@risicle Same for me on NixOS.

**Comment by dotlambda** [comment]
> for me now on master, the `python3.7-CairoSVG-2.3.0` dependency fails its `isort-check` test

fixed in bc14eedcb1766b91ca29ba6adbd62dca2102d17c

@elohmeier Once you rebase your branch on master, you'll see that weasyprint fails to build with `pytest: command not found` due to https://github.com/NixOS/nixpkgs/pull/57377#discussion_r264424912.

**Comment by risicle** [comment]
@dotlambda : nice - was working on a fix myself - will drop that commit - but I'm also finding `pythonPackages.qiskit` and `pythonPackages.pygal` which depend on _it_ are failing to build now. Should have a PR soon.

File: pkgs/development/python-modules/weasyprint/default.nix (hunk 5 above)
Comment (elohmeier) [reply pkgs/development/python-modules/weasyprint/default.nix]: done

Comment (elohmeier) [reply pkgs/development/python-modules/weasyprint/default.nix]: done

**Comment by elohmeier** [comment]
@dotlambda thanks for your support! I've rebased and was able to build successfully now.

File: pkgs/development/python-modules/weasyprint/default.nix [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        --replace '[tool:pytest]' '[tool:pytest]\nflake8-ignore = E501'
+  '';
+
+  checkInputs = [ pytest pytestrunner pytest-isort pytest-flake8 pytestcov ];
+
+  FONTCONFIG_FILE = "${fontconfig.out}/etc/fonts/fonts.conf";
+  FONTCONFIG_PATH = "${fontconfig.out}/etc/fonts/";
+
+  buildInputs = [ glibcLocales ];
+  LC_ALL="en_US.UTF-8";
```
Comment (dotlambda) [inline pkgs/development/python-modules/weasyprint/default.nix]: These two lines are no longer necessary.

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python2.pkgs.weasyprint python3.pkgs.weasyprint

File: pkgs/development/python-modules/weasyprint/default.nix (hunk 6 above)
Comment (elohmeier) [reply pkgs/development/python-modules/weasyprint/default.nix]: Thanks for your feedback, I've removed them!

File: pkgs/top-level/python-packages.nix [hunk 7]
```diff
@@ -4377,6 +4377,8 @@ in {
 
   uritemplate = callPackage ../development/python-modules/uritemplate { };
 
+  weasyprint = callPackage ../development/python-modules/weasyprint { };
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: Sorry I didn't find this earlier. This file should be ordered somewhat alphabetically.

Comment (elohmeier) [reply pkgs/top-level/python-packages.nix]: Thanks for taking a look, I've fixed that!

**Comment by FRidh** [comment]
@GrahamcOfBorg build python2.pkgs.weasyprint python3.pkgs.weasyprint

**Comment by risicle** [comment]
Also builds & tests pass for me on macos 10.13.

**Comment by elohmeier** [comment]
I think there are no open tasks for this package, can we get this merged?

**Pull request merged**